File: leetcode/Python/reverse-only-letters.py
class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        def _isalphabet(c):
            if (c>='a' and c<='z') or (c>='A' and c<='Z'):
                return True
            return False
        sol = list(s)
        for i in range(len(sol)):
            if _isalphabet(sol[i]):
                sol[i]=0
        i, j = 0, len(s)-1
        while i<len(s) and j>=0:
            while i<len(s) and sol[i]!=0:
                i+=1
            while j>=0 and not _isalphabet(s[j]):
                j-=1
            if i<len(s) and j>=0:
                sol[i]=s[j]
                i, j = i+1, j-1
        return ''.join(sol)

File: leetcode/C++/reverse-only-letters.cpp
class Solution {
public:
    bool isalphabet(char &c){
        if((c>='a'&&c<='z') || (c>='A'&&c<='Z')) return true;
        return false;
    }
    string reverseOnlyLetters(string s) {
        string sol=s;
        for(char &c : sol) if(isalphabet(c)) c=0;
        for(int i=0,j=s.length()-1; i<s.length() && j>=0;){
            while(i<sol.length() && sol[i]!=0) ++i;
            while(j>=0 && !isalphabet(s[j])) --j;
            if(i<s.length() && j>=0) sol[i++]=s[j--];
        }
        return sol;
    }
};